README
NAME
    App::CPAN::Mini::Visit - explore each distribution in a minicpan
    repository

VERSION
    This documentation describes 0.007.

SYNOPSIS
         #!/usr/bin/perl
         use App::CPAN::Mini::Visit;
         exit App::CPAN::Mini::Visit->run;

DESCRIPTION
    This module contains the guts of the visitcpan program. See
    documentation of that program for details on features and command line
    options.

USAGE
  `run()'
         exit App::CPAN::Mini::Visit->run();

    Executes the program, processing command line arguments and traversing a
    minicpan repository. Returns an exit code.

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    http://rt.cpan.org/Dist/Display.html?Queue=App-CPAN-Mini-Visit

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   CPAN::Mini

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2008-2009 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from
    http:E<sol>E<sol>www.apache.orgE<sol>licensesE<sol>LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for App-CPAN-Mini-Visit

{{$NEXT}}

    [CHANGED]

    - Actually fix the repo URL this time

0.007     2014-02-27 11:31:54-05:00 America/New_York

    [CHANGED]

    - Updated metadata, repo, and bugtracker

    - Switched distribution to Dist::Zilla management

0.006 Sat Oct 24 22:57:07 EDT 2009
  Fixed:
    - Fixed bug in testing [Ricardo Signes]

0.005 Mon Feb 23 20:35:53 EST 2009
  Fixed:
    - removed Capture::Tiny dependency 
    - removed unused version.pm prereq

0.004 Mon Feb 23 08:48:49 EST 2009
  Added:
    - "-e" option to act like "perl -e"
    - "--output" option to capture output in a file
    - examples in visitcpan POD

  Changed:
    - returned to old-style version numbers

v0.3.1 Fri Dec 12 19:42:21 EST 2008
    - fixed: usage help text

v0.3.0 Fri Dec 12 15:43:53 EST 2008
    - added: "--quiet" option to suppress warnings and tar errors

v0.2.0 Fri Dec 12 01:40:23 EST 2008
    - added: "--append" option to provide path or dist name as an argument

v0.1.0 Fri Nov  7 15:06:55 EST 2008
    - initial release


Build.PL
# Copyright (c) 2008 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use 5.006;
use strict;
use warnings;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new( 
    module_name => 'App::CPAN::Mini::Visit',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
      'perl' => 5.006,
      'Archive::Extract' => 0.28, 
      'CPAN::Mini' => 0.572,
      'Exception::Class::TryCatch' => 1.12,
      'File::Basename' => 0,
      'File::Find' => 0,
      'File::pushd' => 1.00,
      'Getopt::Lucid' => 0.16,
      'Path::Class' => 0.16,
      'Pod::Usage' => 1.35, 
    },
    build_requires => {
      'File::Temp' => 0.20, 
      'IO::CaptureOutput' => 1.0801,
      'IO::File' => 0,
      'Test::More' => 0.62, 
    },
    meta_add            => { 
        no_index => { 
            directory => [ qw/ examples inc t /],
        } 
    },
);
$build->create_build_script;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3603
require 5.010;
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'App::CPAN::Mini::Visit',
          'VERSION_FROM' => 'lib/App/CPAN/Mini/Visit.pm',
          'PREREQ_PM' => {
                           'Archive::Extract' => '0.28',
                           'CPAN::Mini' => '0.572',
                           'Exception::Class::TryCatch' => '1.12',
                           'File::Basename' => 0,
                           'File::Find' => 0,
                           'File::Temp' => '0.2',
                           'File::pushd' => '1',
                           'Getopt::Lucid' => '0.16',
                           'IO::CaptureOutput' => '1.0801',
                           'IO::File' => 0,
                           'Path::Class' => '0.16',
                           'Pod::Usage' => '1.35',
                           'Test::More' => '0.62'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [
                           'bin/visitcpan'
                         ],
          'PL_FILES' => {}
        )
;


lib/App/CPAN/Mini/Visit.pm
use 5.006;
use strict;
use warnings;

package App::CPAN::Mini::Visit;
# ABSTRACT: explore each distribution in a minicpan repository
# VERSION

use CPAN::Mini 0.572 ();
use Exception::Class::TryCatch 1.12 qw/ try catch /;
use File::Basename qw/ basename /;
use File::Find qw/ find /;
use File::pushd qw/ tempd /;
use Path::Class qw/ dir file /;
use Getopt::Lucid 0.16 qw/ :all /;
use Pod::Usage 1.35 qw/ pod2usage /;

use Archive::Extract 0.28 ();

my @option_spec = (
    Switch("help|h"), Switch("version|V"),
    Switch("quiet|q"), Param( "append|a", qr/(?:^$|(?:^path|dist$))/ )->default(''),
    Param("e|E"),      Param("minicpan|m"),
    Param("output|o"),
);

sub run {
    my ( $self, @args ) = @_;

    # get command line options
    my $opt = try eval { Getopt::Lucid->getopt( \@option_spec, \@args ) };
    for (catch) {
        if ( $_->isa('Getopt::Lucid::Exception::ARGV') ) {
            print "$_\n";
            # usage stuff
            return 1;
        }
        else {
            die $_;
        }
    }

    # handle "help" and "version" options
    return _exit_usage()   if $opt->get_help;
    return _exit_version() if $opt->get_version;

    # Set Archive::Extract globals
    # if quiet suppress warnings from Archive::Tar, etc.
    local $Archive::Extract::DEBUG      = 0;
    local $Archive::Extract::PREFER_BIN = 1;
    local $Archive::Extract::WARN       = $opt->get_quiet ? 0 : 1;

    # if -e/-E, then prepend to command
    if ( $opt->get_e ) {
        unshift @args, $^X, ( $^X > 5.009 ? '-E' : '-e' ), $opt->get_e;
    }

    # locate minicpan directory
    if ( !$opt->get_minicpan ) {
        my %config = CPAN::Mini->read_config;
        if ( $config{local} ) {
            $opt->merge_defaults( { minicpan => $config{local} } );
        }
    }

    # confirm minicpan directory that looks like minicpan
    return _exit_no_minicpan() if !$opt->get_minicpan;
    return _exit_bad_minicpan( $opt->get_minicpan ) if !-d $opt->get_minicpan;

    my $id_dir = dir( $opt->get_minicpan, qw/authors id/ );
    return _exit_bad_minicpan( $opt->get_minicpan ) if !-d $id_dir;

    # process all distribution tarballs in authors/id/...
    my $archive_re = qr{\.(?:tar\.(?:bz2|gz|Z)|t(?:gz|bz)|zip|pm\.gz)$}i;

    my $minicpan = dir( $opt->get_minicpan )->absolute;

    # save output by redirecting STDOUT if requested
    my ( $out_fh, $orig_stdout );
    if ( $opt->get_output ) {
        open $out_fh, ">", $opt->get_output;
        open $orig_stdout, "<&=STDOUT";
        open STDOUT, ">&=" . fileno $out_fh;
    }

    find(
        {
            no_chdir   => 1,
            follow     => 0,
            preprocess => sub { my @files = sort @_; return @files },
            wanted     => sub {
                return unless /$archive_re/;
                # run code if program/args given otherwise print name
                if (@args) {
                    return if $_ =~ /pm\.gz$/io; # not an archive, just a file
                    my @cmd = @args;
                    if ( $opt->get_append ) {
                        if ( $opt->get_append eq 'dist' ) {
                            my $distname = $_;
                            my $prefix = dir( $minicpan, qw/authors id/ );
                            $distname =~ s{^$prefix[\\/].[\\/]..[\\/]}{};
                            push @cmd, $distname;
                        }
                        else {
                            push @cmd, $_;
                        }
                    }
                    _visit( $_, @cmd );
                }
                else {
                    print "$_\n";
                }
            },
        },
        $minicpan
    );

    # restore STDOUT and close output file
    if ( $opt->get_output ) {
        open STDOUT, ">&=" . fileno $orig_stdout;
        close $out_fh;
    }

    return 0; # exit code
}

sub _exit_no_minicpan {
    print STDERR << "END_NO_MINICPAN";
No minicpan configured.

END_NO_MINICPAN
    return 1;
}

sub _exit_bad_minicpan {
    my ($dir) = @_;
    die "requires directory argument" unless defined $dir;
    print STDERR << "END_BAD_MINICPAN";
Directory '$dir' does not appear to be a CPAN repository.

END_BAD_MINICPAN
    return 1;
}

sub _exit_usage {
    my $exe = basename($0);
    print STDERR << "END_USAGE";
Usage:
  $exe [OPTIONS] [PROGRAM]

  $exe [OPTIONS] -- [PROGRAM] [ARGS]

Options:

 --append|-a        --append=dist -> append distname after ARGS
                    --append=path -> append tarball path after ARGS

 -e|-E              run next argument via 'perl -E'
 
 --help|-h          this usage info 

 --minicpan|-m      directory of a minicpan (defaults to local minicpan 
                    from CPAN::Mini config file)

 --output|-o        file to save output instead of sending to terminal

 --quiet|-q         silence warnings and suppress STDERR from tar

 --version|-V       $exe program version

 --                 indicates the end of options for $exe


END_USAGE
    return 1;
}

sub _exit_version {
    print STDERR basename($0) . ": " . ( __PACKAGE__->VERSION() || 0 ) . "\n";
    return 1;
}

sub _visit {
    my ( $archive, @cmd_line ) = @_;

    my $tempd = tempd;

    my $ae = Archive::Extract->new( archive => $archive );

    my $olderr;

    # stderr > /dev/null if quiet
    if ( !$Archive::Extract::WARN ) {
        open $olderr, ">&STDERR";
        open STDERR, ">", File::Spec->devnull;
    }

    my $extract_ok = $ae->extract;

    # restore stderr if quiet
    if ( !$Archive::Extract::WARN ) {
        open STDERR, ">&", $olderr;
        close $olderr;
    }

    if ( !$extract_ok ) {
        warn "Couldn't extract '$archive'\n" if $Archive::Extract::WARN;
        return;
    }

    # most distributions unpack a single directory that we must enter
    # but some behave poorly and unpack to the current directory
    my @children = dir()->children;
    if ( @children == 1 && -d $children[0] ) {
        chdir $children[0];
    }

    # execute command
    my $rc = system(@cmd_line);
    if ( $rc == -1 ) {
        warn "Error running '@cmd_line': $!\n";
    }

    return;
}

1;

__END__

=head1 SYNOPSIS

    #!/usr/bin/perl
    use App::CPAN::Mini::Visit;
    exit App::CPAN::Mini::Visit->run;

=head1 DESCRIPTION

This module contains the guts of the L<visitcpan> program.  See documentation of
that program for details on features and command line options.

=head1 USAGE

=head2 C<run>

    exit App::CPAN::Mini::Visit->run();

Executes the program, processing command line arguments and traversing
a minicpan repository.  Returns an exit code.

=head1 SEE ALSO

=for :list
* L<CPAN::Mini>

=cut
